include/dsn/internal/network.h

```diff
@@ -284,14 +284,17 @@ namespace dsn {
         dlink                              _messages;        
         volatile session_state             _connect_state;
         uint64_t                           _message_sent;
+        // ]
+
+        ::dsn::utils::ex_lock_nr           _delay_lock; // [
         int                                _delay_server_receive_ms;
         // ]
     };
 
     // --------- inline implementation --------------
     inline void rpc_session::delay_recv(int delay_ms)
     {
-        utils::auto_lock<utils::ex_lock_nr> l(_lock);
+        utils::auto_lock<utils::ex_lock_nr> l(_delay_lock);
```

Comment:
Yes, good for performance. I did not do that because I'm not sure the throttling mechanism is good enough, which may be changed later. Anyway, feel free to change that to atomic ops.